File: 542-01-matrix/01-matrix.py
class Solution:
    
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        m, n = len(mat), len(mat[0])
        visited = [[False]*n for i in range(m)]
        res = [[0]*n for i in range(m)]
        q = []
        delRow = [-1,0,1,0]
        delCol = [0,1,0,-1]
        # comment = ['t','r','b','l']

        for i in range(m):
            for j in range(n):
                if mat[i][j]==0:
                    visited[i][j]=True
                    q.append((i,j,0))
                    # print('+q:',q[-1])
        
        while len(q)>0:
            i, j, dist = q.pop(0)
            for d in range(4):
                nrow = delRow[d]+i
                ncol = delCol[d]+j
                if nrow>=0 and nrow<m and ncol>=0 and ncol<n and not visited[nrow][ncol]:
                    # print('bfs:',comment[d], i,j,' = ',dist+1)
                    visited[nrow][ncol] = True
                    res[nrow][ncol] = dist+1
                    q.append((nrow,ncol,dist+1))
        
        return res




    

File: 542-01-matrix/01-matrix.java
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;
        boolean[][] visited = new boolean[m][n];
        int[][] res = new int[m][n];
        Queue<List<Integer>> q = new ArrayDeque<>();

        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(mat[i][j]==0){
                    visited[i][j]=true;
                    List<Integer> l = new ArrayList<>();
                    l.add(i);
                    l.add(j);
                    l.add(0);
                    q.add(l);
                }
            }
        }

        int[] di = new int[]{-1,0,1,0};
        int[] dj = new int[]{0,1,0,-1};
        while(q.iterator().hasNext()){
            List<Integer> tuple = q.poll();
            int i = tuple.get(0);
            int j = tuple.get(1);
            int dist = tuple.get(2);

            for(int d=0;d<4;d++){
                int nrow = i+di[d];
                int ncol = j+dj[d];

                if(inRange(nrow,0,m) && inRange(ncol,0,n) && !visited[nrow][ncol]){
                    res[nrow][ncol] = dist+1;
                    List<Integer> l = new ArrayList<>();
                    l.add(nrow);
                    l.add(ncol);
                    l.add(dist+1);
                    q.add(l);
                    visited[nrow][ncol] = true;
                }
            }
        }

        return res;

    }

    public boolean inRange(int n,int left, int right){
        return n>=left && n<right;
    }
}